File: assn3/qsort.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cstdlib>

using namespace std;

//not the best way but works for out purpose
int randomInteger(int min, int max) {
	int range = max - min + 1;
	int num = rand() % range + min;

	return num;

}

//print the arrays
void printArray(vector<int> arr, vector<int> arr2) {
	cout << "array: " << arr.size() << endl;
	for (int x = 0; x<sizeof(arr); x++) {
		cout << arr[x] << endl;
	}

	cout << "output: " << endl;
	for (int x = 0; x<sizeof(arr2); x++) {
		cout << arr2[x] << endl;
	}
}


void swap(vector<int> arr, int i, int j)
{
	int temp = arr[i];
	arr[i] = arr[j];
	arr[j] = temp;
}

void swap(int* a, int* b)
{
	int t = *a;
	*a = *b;
	*b = t;
}

int partition(vector<int> & arr, int low, int high)
{
	int pivot = arr[high];
	int i = (low - 1);  //the smaller partition is till low -1

	for (int j = low; j <= high - 1; j++)
	{
		//check for the pibot position relative to smaller element
		if (arr[j] <= pivot)
		{
			i++;    // increment index of smaller element
			swap(&arr[i], &arr[j]);
		}
	}
	swap(&arr[i + 1], &arr[high]);	//swap the elements
	return (i + 1);

}
void quickSort(vector<int> & arr, int low, int high)
{
	if (low < high)
	{
		int pi = partition(arr, low, high);	//partioning index

		quickSort(arr, low, pi - 1);	//sort 1st half
		quickSort(arr, pi + 1, high);	//sort the other half
	}
}




int main()
{
	//example test code
	/*vector<int> a{ 1, 3, 2, 6, 4, 5, 8, 7, 9, 0 };
	quickSort(a, 0, a.size() - 1);
	for (int x = 0; x < a.size(); x++) {

	cout<<a[x]<<endl;
	}*/


	int noOfElements = 0;
	cout << "CPP Quicksort" << endl;
	cout << "Enter Size of Array:" << endl;
	cin >> noOfElements;
	cout << "no. of elem: " << noOfElements << endl;

	vector<int> newArr;
	vector<int> output;

	for (int i = 0; i < noOfElements; i++) {
		cout << "Enter Value for A[" << i << "]" << endl;
		int inputFromUser = 0;
		cin >> inputFromUser;
		newArr.push_back(inputFromUser);	//add user input integer to the vector
	}

	output = newArr;

	quickSort(output, 0, output.size() - 1);


	cout << "Input: [";
	for (int x = 0; x< newArr.size(); x++) {
		cout << newArr[x];
		cout << ", ";
	}
	cout << "]" << endl;

	cout << "Output: [";
	for (int x = 0; x< output.size(); x++) {
		cout << output[x];
		cout << ", ";
	}
	cout << "]" << endl;

	output.clear();
	newArr.clear();

	return 0;
}

File: assn3/qsort.java
import java.util.Scanner;

class Qsort
{
    int partition(int arr[], int low, int high)
    {
        int pivot = arr[high];
        int i = (low-1);    //the smaller partition is till low -1
        for (int j=low; j<high; j++)
        {
            //check for the pibot position relative to smaller element
            if (arr[j] <= pivot)
            {
                i++;    // increment index of smaller element
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        //swap the elements
        int temp = arr[i+1];
        arr[i+1] = arr[high];
        arr[high] = temp;

        return i+1;
    }

    //sort function
    void sort(int arr[], int low, int high)
    {
        if (low < high)
        {
            //similar to the c++ function, pi is the partition index
            int pi = partition(arr, low, high);

            sort(arr, low, pi-1);   //sort 1st half
            sort(arr, pi+1, high);  //sort the other half
        }
    }

    // The main execution function
    public static void main(String args[])
    {
        Scanner reader = new Scanner(System.in);
        int noOfElements = 0;
        System.out.print("Java Quicksort");
        System.out.print("\n");
        System.out.print("Enter Size of Array:");
        System.out.print("\n");
        noOfElements = reader.nextInt();

        int[] arr = new int[noOfElements];

        for (int i = 0; i < noOfElements; i++)
        {
            System.out.print("Enter Value for A[");
            System.out.print(i);
            System.out.print("]");
            System.out.print("\n");
            int inputFromUser = 0;
            inputFromUser = reader.nextInt();   //add user input integer to the array
            arr[i] = inputFromUser;
        }

        //print out the user input array
        System.out.print("Input: [");
        for (int x = 0; x < noOfElements; x++)
        {
            System.out.print(arr[x]);
            if(x!=noOfElements-1) {
                System.out.print(", ");
            }
        }
        System.out.print("]");
        System.out.print("\n");

        //create an object of the class
        Qsort obj = new Qsort();
        obj.sort(arr, 0, ((arr.length)-1));     //use the sort function on the array

        //print out the actual array
        System.out.print("Output: [");
        for (int x = 0; x < noOfElements; x++)
        {
            System.out.print(arr[x]);
            if(x!=noOfElements-1) {
                System.out.print(", ");
            }
        }
        System.out.print("]");
        System.out.print("\n");
    }
} 
